curs/curs-18-19-linking/README.md

```diff
@@ -0,0 +1,436 @@
+```
+[..]/01-one-file$ file one.o
+one.o: ELF 32-bit LSB relocatable, Intel 80386, version 1 (SYSV), with debug_info, not stripped
+
+[..]/01-one-file$ readelf -h one.o
+ELF Header:
+  [..]
+  Type:                              REL (Relocatable file)
+  [..]
+```
+
+Linkerul se ocupă de relocarea referințelor la simboluri.
+Atunci când creează fișierul executabil, urmărește referințele la simboluri și le înlocuiește cu adresele efective ale simbolurilor.
+Pentru a realiza relocarea, adică înlocuirea referințelor, linkerul trebuie să știe unde se găsesc aceste referințe.
+Pentru aceasta fișierele obiect relocabile conțin **tabele de relocare** (*relocation tables*).
+
+Folosim `readelf` pentru a obține tabelele de relocare ale fișierului obiect relocabil `one.o`:
+```
+[..]/01-one-file$ readelf -r one.o
+
+Relocation section '.rel.text' at offset 0x400 contains 4 entries:
+ Offset     Info    Type            Sym.Value  Sym. Name
+00000004  00000d01 R_386_32          00000000   num_items
+0000000c  00000d01 R_386_32          00000000   num_items
+00000018  00000d01 R_386_32          00000000   num_items
+00000021  00000e02 R_386_PC32        00000000   increment
+[...]
+```
+Am selectat mai sus doar partea de interes pentru noi, adică secțiunea `.rel.text`.
+Celelalte informații sunt legate de debugging sau nu sunt relevante.
+
+Observăm în tabela de relocare 4 intrări: 3 referă simbolul `num_items`, iar una la simbolul `increment`.
+Relevantă este coloana `offset` unde este indicată referința la simbol, referință ce trebuie înlocuită de linker.
+Prin parcurgerea secțiunii de relocare `.rel.text`, linkerul ia următoarele decizii:
+* la offsetul `0x04` față de începutul secțiunii `.text` trebuie să înlocuiască referința cu adresa simbolului `num_items`
+* la offsetul `0x0c` față de începutul secțiunii `.text` trebuie să înlocuiască referința cu adresa simbolului `num_items`
+* la offsetul `0x18` față de începutul secțiunii `.text` trebuie să înlocuiască referința cu adresa simbolului `num_items`
+* la offsetul `0x21` față de începutul secțiunii `.text` trebuie să înlocuiască referința cu adresa simbolului `increment`
+
+În dezasamblarea fișierului obiectiv `one.o`, observăm că exact acelea sunt offseturile unde se găsesc referințele la simboluri:
+```
+; one.o
+   3:   a1 00 00 00 00          mov    eax,ds:0x0
+   [...]
+   b:   a3 00 00 00 00          mov    ds:0x0,eax
+   [...]
+  16:   c7 05 00 00 00 00 05    mov    DWORD PTR ds:0x0,0x5
+  1d:   00 00 00
+  [...]
+  20:   e8 fc ff ff ff          call   21 <main+0xe>
+  [...]
+```
+La offseturile `0x04`, `0x0c`, `0x18` se găsesc referințe la simbolul `num_items`.
+Necunoscându-se adresa simbolului `num_items` referințele sunt acum marcate cu `0x00000000`.
+
+Similar, la offsetul `0x21` se găsește referința la simbolul `increment`.
+La fel, necunoscându-se adresa simbolului `num_items, este marcată cu un placeholder.
```

Comment:
Lipseste backtickul final de la `num_items`.
"aceasta este marcată" in loc de "este marcată". Se simte nevoia unui subiect explicit acolo.

Reply:
Probabil ar fi bine să fie menționat explicit "referința este marcată". Precizarea "aceasta" ca subiect nu ajută înțelegerea propoziției.